src/metpy/calc/tools.py

```diff
@@ -950,6 +980,219 @@ def wrapper(f, **kwargs):
+                                                                           'longitude')
+                    scale_lat = latitude.metpy.unit_array
+                    scale_lon = longitude.metpy.unit_array
+                    calculate_scales = True
+                except AttributeError:
+                    # Fall back to basic cartesian calculation if we don't have a CRS or we are
+                    # unable to get the coordinates needed for map factor calculation (either
+                    # exiting lat/lon or lat/lon computed from y/x)
+                    calculate_scales = False
+            elif latitude is not None and longitude is not None:
+                try:
+                    proj = Proj(crs)
+                    calculate_scales = True
+                except Exception as e:
+                    # Whoops, intended to use
+                    raise ValueError(
+                        'Latitude and longitude arguments provided so as to make '
+                        'calculation projection-correct, however, projection CRS is '
+                        'missing or invalid.'
+                    ) from e
+            else:
+                calculate_scales = False
+
+            # Do we have everything we need to sensibly calculate the scale arrays?
+            if calculate_scales:
+                scale_lat = scale_lat.squeeze().m_as('degrees')
+                scale_lon = scale_lon.squeeze().m_as('degrees')
+                if scale_lat.ndim == 1 and scale_lon.ndim == 1:
+                    scale_lon, scale_lat = np.meshgrid(scale_lon, scale_lat)
+                elif scale_lat.ndim != 2 or scale_lon.ndim != 2:
+                    raise ValueError('Latitude and longitude must be either 1D or 2D.')
+                factors = proj.get_factors(scale_lon, scale_lat)
+                p_scale = factors.parallel_scale
+                m_scale = factors.meridional_scale
+
+                if grid_prototype is not None:
+                    # Find appropriate coordinates to assign to the scales from the original
+                    # lat/lon DataArrays. Checking size eliminates the metpy_crs coordinate
+                    # well as any spurious coordinates hanging around in subsets due to
+                    # pydata/xarray#7350.
+                    coords = {c: v for c, v in {**latitude.coords, **longitude.coords}.items()
+                              if v.size > 1}
+
+                    # Because of the uses of these arrays in derivatives alongside the data
+                    # fields, they need to have the same number and order of dimensions as
+                    # the data field.
+                    p_scale = xr.DataArray(p_scale,
+                                           coords=coords).broadcast_like(grid_prototype)
+                    m_scale = xr.DataArray(m_scale,
+                                           coords=coords).broadcast_like(grid_prototype)
```

Comment:
This would also require skipping the warning of

https://github.com/dcamron/MetPy/blob/c9fd2612513448bc6e0d1bb84ae2e411d414ffc3/src/metpy/xarray.py#L1293-L1301

when `arg_name` is one of our scale factors (as this is really only a warning for users, it's totally fine to drop down to `Variable` in the internals).

Reply:
Doesn't look like `broadcast` likes being given a `Variable`:
```pytb
src/metpy/calc/tools.py:1176: in wrapper
    return func(*bound_args.args, **bound_args.kwargs)
src/metpy/xarray.py:1287: in wrapper
    broadcasted_args = xr.broadcast(
../../miniconda3/envs/py311/lib/python3.11/site-packages/xarray/core/alignment.py:1049: in broadcast
    args = align(*args, join="outer", copy=False, exclude=exclude)
../../miniconda3/envs/py311/lib/python3.11/site-packages/xarray/core/alignment.py:772: in align
    aligner.align()
../../miniconda3/envs/py311/lib/python3.11/site-packages/xarray/core/alignment.py:556: in align
    self.find_matching_indexes()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <xarray.core.alignment.Aligner object at 0x1a32078d0>

    def find_matching_indexes(self) -> None:
        all_indexes: dict[MatchingIndexKey, list[Index]]
        all_index_vars: dict[MatchingIndexKey, list[dict[Hashable, Variable]]]
        all_indexes_dim_sizes: dict[MatchingIndexKey, dict[Hashable, set]]
        objects_matching_indexes: list[dict[MatchingIndexKey, Index]]
    
        all_indexes = defaultdict(list)
        all_index_vars = defaultdict(list)
        all_indexes_dim_sizes = defaultdict(lambda: defaultdict(set))
        objects_matching_indexes = []
    
        for obj in self.objects:
>           obj_indexes, obj_index_vars = self._normalize_indexes(obj.xindexes)
E           AttributeError: 'Variable' object has no attribute 'xindexes'
```

Reply:
I do like the idea of relying on `preprocess_and_wrap` to handle the broadcasting, though.

Also, we can get less error prone by just passing in the union of coordinates and make sure we pass in the appropriate pair of dimensions from `latitude` and `longitude` rather than filtering on size.

Reply:
Like this:
```python
                    xr_kwargs = {'coords': {**latitude.coords, **longitude.coords},
                                 'dims': (latitude.dims[0], longitude.dims[-1])}
                    p_scale = xr.DataArray(p_scale, **xr_kwargs)
                    m_scale = xr.DataArray(m_scale, **xr_kwargs)
```

Reply:
I guess I was mistaken (and that this test isn't doing the intended thing: https://github.com/Unidata/MetPy/blob/4cc1df04326e9817fc6e42760137fe0e8ac48727/tests/test_xarray.py#L1347-L1369)! But, that is still good that the idea lead to an improved implementation.